tonic/src/transport/server.rs

```diff
@@ -173,6 +175,24 @@ impl Server {
         }
     }
 
+    /// Setup propagation of trace parents requires a function that
+    /// can turn the `grpc-trace-bin` header into a parent span Id
+    pub fn propagate_trace_parents<F>(self, f: F) -> Self
+    where
+        F: Fn(&HeaderValue) -> Option<Id> + Send + Sync + 'static,
+    {
+        self.interceptor_fn(move |svc, req| {
+            if let Some(grpc_trace) = req.headers().get("grpc-trace-bin") {
+                let span_id = f(grpc_trace);
+                let root = span!(parent: span_id, Level::INFO, "request");
+                let _enter = root.enter();
+                svc.call(req)
```

Comment:
This won't be propagated beyond this span, unfortunately. `_enter` would dropped at the end of this scope. You'd need something like: https://github.com/tokio-rs/tracing/blob/master/tracing-tower/src/service_span.rs, which hasn't been updated to the releases of Tower using `std::future::Future`. That might happen this week or next week...?

Reply:
Probably a noob question but is it possible to use `tracing-futures` to do this by `svc.call(req).instrument(root)` ?

Reply:
I'm not 100% sure, but I think the instrumentation needs to be on the Service-level. Give it a shot?

Reply:
> Probably a noob question but is it possible to use `tracing-futures` to do this by `svc.call(req).instrument(root)` ?

Yeah, that will work fine!

(@davidbarsky, this is all that `tracing-tower` does :P)

Reply:
I think allowing users to allow their own span is probably a good idea.